spec/ios_download_strings_files_from_glotpress_spec.rb

```diff
@@ -0,0 +1,181 @@
+          locales: locales_subset,
+          download_dir: tmp_dir,
+          table_basename: tablename,
+          filters: filters
+        }.compact)
+
+        # Assert
+        expect(stub_fr).to have_been_made.once
+        file_fr = File.join(tmp_dir, 'fr.lproj', "#{tablename || 'Localizable'}.strings")
+        expect(File).to exist(file_fr)
+        expect(File.read(file_fr)).to eq('"key" = "fr-copy";')
+
+        expect(stub_zh).to have_been_made.once
+        file_zh = File.join(tmp_dir, 'zh-Hans.lproj', "#{tablename || 'Localizable'}.strings")
+        expect(File).to exist(file_zh)
+        expect(File.read(file_zh)).to eq('"key" = "zh-copy";')
+      end
+    end
+
+    it 'downloads all the locales into the expected directories' do
+      test_gp_download(filters: nil, tablename: nil, expected_gp_params: { 'filters[status]': 'current', format: 'strings' })
+    end
+
+    it 'uses the proper filters when exporting the files from GlotPress' do
+      test_gp_download(
+        filters: { term: 'foo', status: 'review' },
+        tablename: nil,
+        expected_gp_params: { 'filters[term]': 'foo', 'filters[status]': 'review', format: 'strings' }
+      )
+    end
+
+    it 'uses a custom table name for the `.strings` files if provided' do
+      test_gp_download(
+        filters: nil,
+        tablename: 'MyApp',
+        expected_gp_params: { 'filters[status]': 'current', format: 'strings' }
+      )
+    end
+  end
+
+  describe 'error handling' do
+    it 'shows an error if an invalid locale is provided (404)' do
+      Dir.mktmpdir('a8c-release-toolkit-tests-') do |tmp_dir|
+        # Arrange
+        stub = gp_stub(locale: 'unknown-locale', query: { 'filters[status]': 'current', format: 'strings' }).to_return(status: [404, 'Not Found'])
+        error_messages = []
+        allow(FastlaneCore::UI).to receive(:error) { |message| error_messages.append(message) }
+
+        # Act
+        run_described_fastlane_action(
+          project_url: gp_fake_url,
+          locales: { 'unknown-locale': 'Base' },
+          download_dir: tmp_dir
+        )
+
+        # Assert
+        expect(stub).to have_been_made.once
+        file = File.join(tmp_dir, 'Base.lproj', 'Localizable.strings')
+        expect(File).not_to exist(file)
+        expect(error_messages).to eq(['Error downloading locale `unknown-locale` — 404 Not Found'])
+      end
+    end
+
+    it 'shows an error if the file cannot be written in the destination' do
+      # Arrange
+      download_dir = '/these/are/not/the/dirs/you/are/looking/for/'
+
+      # Act
+      act = lambda do
+        run_described_fastlane_action(
+          project_url: gp_fake_url,
+          locales: { 'fr-FR': 'fr' },
+          download_dir: download_dir
+        )
+      end
+
+      # Assert
+      expect { act.call }.to raise_error(FastlaneCore::Interface::FastlaneError, "The parent directory `#{download_dir}` (which contains all the `*.lproj` subdirectories) must already exist")
```

Comment:
This is the exception raised by `UI.user_error!`.

PS: Note that we can't really use `allow(FastlaneCore::UI).to receive(:user_error!) { |m| … }` here because if we catch and stub those, that won't interrupt the action (i.e. the call [here on L9](https://github.com/wordpress-mobile/release-toolkit/pull/331/files#diff-db88a4ce40ab03bd414faa6cafc8fde0008f5fdf43b7db981943fd9ab40fd73fR9) would not raise if stubbed so the rest of the `run` method would continue executing).

_(And we could do `allow(FastlaneCore::UI).to receive(:user_error!) do |e| raise e }` to be sure that it still interrupts the code even when stubbed… but then what would be the point of stubbing the method while the real implementation already raises…)_

Reply:
Btw, wdyt of using a `lambda` like this here (which is basically the Ruby equivalent of an anonymous closure in Swift) so that it allowed me to still keep the `# Act` and `# Assert` separate?

Is it readable and worth it? Or too obscure and you'd instead prefer to avoid the `lambda` and directly use `expect { run_described_fastlane_action(…) }.to …`… even if that means mixing the "Act" and "Assert" steps together and breaking the usual A/A/A structure of the tests just to be able to catch the exception raised by the "act" part?

Reply:
> Btw, wdyt of using a `lambda` like this here...

I think it's a neat trick that doesn't affect the code readability much. Having said that, as much as I like AAA I think it would be okay to break the pattern here to keep the code concise.

Up to your preference.

Reply:
👍  I'll keep it then 🙂